pkg/services/k8s/apiserver/admission/grpc_callout_admission_in_plugin.go

```diff
@@ -0,0 +1,183 @@
+	var finalErr error
+
+	kind := a.GetKind()
+	// quit early if plugin doesn't recognize the kind
+	if kind.Group != "charandas.example.com" || kind.Kind != "TestObject" {
+		return nil
+	}
+
+	wrappedSender := callResourceResponseSenderFunc(func(response *backend.CallResourceResponse) error {
+		// err = errors.New("some validation error")
+		admissionResponse := &admissionV1.AdmissionResponse{}
+		if err := json.Unmarshal(response.Body, &admissionResponse); err != nil {
+			finalErr = errors.New("Admission response from plugin is malformed")
+			return nil
+		}
+
+		if !admissionResponse.Allowed {
+			finalErr = admission.NewForbidden(a, errors.New("could not pass validation performed by plugin"))
+			return nil
+		}
+
+		if isMutating {
+			obj := a.GetObject()
+			var patchedJS []byte
+			jsonSerializer := runtimeJson.NewSerializer(runtimeJson.DefaultMetaFactory, o.GetObjectCreater(), o.GetObjectTyper(), false)
+			switch *admissionResponse.PatchType {
+			// VerifyAdmissionResponse normalizes to v1 patch types, regardless of the AdmissionReview version used
+			case admissionV1.PatchTypeJSONPatch:
+				objJS, err := runtime.Encode(jsonSerializer, obj)
+				if err != nil {
+					finalErr = admission.NewForbidden(a, errors.New("Could not serialize object to JSON for patching"))
+				} else {
+					patchObj, _ := jsonpatch.DecodePatch(admissionResponse.Patch)
+					patchedJS, _ = patchObj.Apply(objJS)
+					jsonSerializer.Decode(patchedJS, &schema.GroupVersionKind{
+						Group:   "charandas.example.com",
+						Version: "v1",
+						Kind:    "TestObject",
+					}, obj)
+
+				}
+
+			}
+		}
+
+		return nil
+	})
+
+	userInfo := a.GetUserInfo()
+	admissionRequest := &admissionV1.AdmissionRequest{
+		UID:             "0",
+		Kind:            v1.GroupVersionKind(a.GetKind()),
+		Resource:        v1.GroupVersionResource(a.GetResource()),
+		SubResource:     a.GetSubresource(),
+		RequestKind:     nil,
+		RequestResource: nil,
+		Name:            a.GetName(),
+		Namespace:       a.GetNamespace(),
+		Operation:       admissionV1.Operation(a.GetOperation()),
+		UserInfo: authenticationV1.UserInfo{
+			Username: userInfo.GetName(),
+			Groups:   userInfo.GetGroups(),
+			UID:      userInfo.GetUID(),
+		},
+		Object: runtime.RawExtension{
+			Object: a.GetObject(),
+		},
+		OldObject: runtime.RawExtension{
+			Object: a.GetOldObject(),
+		},
+		DryRun:  nil,
+		Options: runtime.RawExtension{},
+	}
+
+	admissionRequestBody, _ := json.Marshal(admissionRequest)
+	path := "/k8s/admission/validation"
+
+	if isMutating {
+		path = "/k8s/admission/mutation"
+	}
+
+	finalErr = g.pluginsClient.CallResource(ctx, &backend.CallResourceRequest{
+		Path:   path,
+		Method: "POST",
+		PluginContext: backend.PluginContext{
+			OrgID:                      0,
+			PluginID:                   "charandas-callbackadmissionexample-app",
```

Comment:
@sdboyer this is where I hard coded the plugin id where the admission decision was forwarded to.

Ideally, there would be a way for Grafana to also tell which kind is associated to which plugin.

Also, recognizing that this is not the most preferred approach and when we do go for the aggregated server approach for each plugin, the registration for the specific APIService will include the kind information for the plugin and the mapping will become declarative.

Still thinking about what to make this APIService construct out of i.e. whether there should be a url property on the GrafanaResource before being registered in Grafana.

Reply:
ahh! ok, missed this bit. Thank you!

Now this is probably just my lack of plugin knowledge, but i would have imagined that interactions like this are generally mediated through the plugin manager (perhaps just when it comes to getting one of these `PluginContext`s?). Otherwise i don't know how interactions like this would be able to abstract over local vs. remote/shared plugins.

If that is the expected eventual path, then good, two birds one stone, because...

> Ideally, there would be a way for Grafana to also tell which kind is associated to which plugin.

#77184 shows current trajectory - providing kinds is tightly bound to plugin loading/installation, so there'll be a very clear ownership relationship between kinds and plugins.

But if direct connections through a client like this are the generally expected approach long-term, then unfortunately i don't know enough to say anything about whether we're heading towards convergence or divergence 😢

Reply:
Direct connections are not a decided upon approach. I was merely trying to get something working over gRPC and this worked out quickly.